README.txt
Package description
*******************

.. contents::

.. Note!
   -----
   
   - code repository
   - bug tracker
   - questions/comments feedback mail


- Code repository: http://svn.syslab.com/svn/OSHA/osha.policy
- Questions and comments to info (at) syslab (dot) com
- Report bugs at http://products.syslab.com/products/osha.policy/issues


JavaScript Registry
===================

The JS registry: (portal_javascripts) is cleared and re-imported from osha/policy/profiles/default/osha-jsregistry.xml by the resetJSRegistry setuphandler. Please note that all JavaScript should be managed in this file to avoid trouble with re-ordering etc. caused by installed JS via other profiles.


VDEX Tools
==========
This egg provides a number of shell scripts that shall support people in
updating vdex definitions. They are meant for developers and have not a lot
of error handling.

cmd_parser
    Tool to parse some excel file for generating command lists for the vdex
    updater. This is error prone and should be avoided

vdex_updater
    Tool to update a vdex file. You should provide a file with a command list
    and the vdex file. The command list is a little dsl that is described below.
    The output will always be stored in the file ./out.xml

xmltoxsl
    This works only with thesaurus files, and gives some statistical information
    about the vdex file. Its output will always be ./out.csv

ordervdex
    This tool takes a vdex file and orders the element according to id.
    It overrides the old xml file!

insertTranslations
    This tool takes the vdex file and then a list of translations provided
    as xls as in the form given out from xltoxsl.
    It assumes that the language id is in the filename in [27:29].
    It inserte the provided translations.
    The outputfile is always new.vdex.

VDEX UPDATER DSL
================
A line is one command.
A line consists of fields, separated by pipe characters.
The first field is the command.

c = Create
    The next fields contain the new id, the parent id, a caption
    and a description that is currently ignored.
    The new term will be added below the last child if the parent term
    those termIdentifier is lower than the new id. Thus, an ordered
    tree will still be ordered.

u = Update
    The next fields contain the id, the caption and description. The
    description is currently ignored.

m = Move
    The next fields contain the old id, the new id and the new father id.
    The Term gets removed from the old place and added below the father.
    Where it gets added is calculated with the same rules as the Create
    command. All commands have another parameter now, the last one, must
    be keep_names or something else. This is only relevant if you copy
    your term over to an existing term. with keep_names, in any combination
    of upper and lower case letter, we will not do anything. With another
    value, like, overwrite_names, we will override the target term name
    and its translations.

d = Delete
    The next field contains the id of the term to be deleted.


Testing
=======

osha.policy contains tests for the package itself and other dependencies
in the osha namespace (e.g. osha.theme) - these tests might move to their
respective packages in the future.

To run all the tests for osha.policy::

    $ bin/test-osha

To run a specific test from osha.policy::

    $ bin/test-osha -t test_oshaview

Running this command might take quite a lot of time, because there are
some very time consuming tests. If you run the tests frequently it is
advisable to use an alternative test runner, which runs the tests
without all the expensive tests.

To run a test runner for osha.policy without the time consuming tests::

    $ bin/test-osha-quick

Note that if you want to use the '-t' switch, just specify the filter 
immediately after the test runner script, without the '-t' (because
we already use the -t switch in the test runner to exclude the slow
tests)::

    $ bin/test-osha-quick test_oshaview

To run the tests for the other packages which are dependencies of
osha.policy::

    $ bin/test-deps

For other test options please consult the test runner help::

    $ bin/test --help


Credits
=======

Copyright European Agency for Health and Safety at Work and Syslab.com
GmbH.

osha.policy development was funded by the European Agency for Health
and Safety at Work.


License
=======

osha.policy is licensed under the GNU Lesser Generic Public License,
version 2 or later and EUPL version 1.1 only. The complete license
texts can be found in docs/LICENSE.GPL and docs/LICENSE.EUPL.


CHANGES.txt
osha.policy Changelog
=====================

1.4.40 (unreleased)
-------------------

- Nothing changed yet.


1.4.39 (2014-10-24)
-------------------

- fixed start and end dates in jsonfeed, refs #10803

1.4.38 (2014-08-18)
-------------------

- Don't offer News & Events in jsonfeed if they are outdated and expired #10478
  [pysailor]

1.4.37 (2014-06-17)
-------------------

- Nothing changed yet.


1.4.36 (2014-04-29)
-------------------

- prepare translations for new napo film #9738
- changes required for revamped OSHMail #9323 [deroiste, pysailor]


1.4.35 (2014-03-24)
-------------------

- Made Subject language independent again #9706
- Subject computed based on Subcategory also for News #9694
  [pysailor]


1.4.34 (2013-12-06)
-------------------

- override the default TranslationFactory from LinguaPlone, so that when a
  new translation is created, the language independent fields are only copied to
  the new translation, not to ALL translations
  This fixes a massive performance weakness. Refs #9203 [pysailor]

1.4.33 (2013-12-04)
-------------------

- Improved the script to cleanup context #7803 [pysailor]

1.4.32 (2013-11-18)
-------------------

- Added JSON query feed for providing content to the new campaign site,
  #8937 [regebro]

1.4.31 (2013-10-25)
-------------------

- #8680 Email blog author when a comment is added to their article
  [deroiste]
- #8938 Add captcha for p.a.discussion [deroiste]


1.4.30 (2013-09-23)
-------------------

- Updated jsregistry with scripts needed for plone.app.discussion [pysailor]

1.4.29 (2013-08-30)
-------------------

- Added osha.applicationform as dependency again [pysailor]

1.4.28 (2013-08-21)
-------------------

- fix #8008: updated link checker (lc_maintenance) view to use the
  reference_catalog instead of normal catalog to fetch the objects. This
  improves perfomance when used together with LDAP. Also improved perfomance
  of writing to postgres db by executing and commiting insert statements in
  batches of 1000. Cleaned up and removed old unused code.
  [jcerjak]
- #7716 comment notifications: add upgrade step to migrate
  p.a.discussion items [deroiste]
- Added Products.LoginLockout and Products.PasswordStrength to requirements,
  plus upgrade-step to install and configure them.
  #7594, #7595 [regebro]


1.4.27 (2013-07-23)
-------------------

- rewrote search_solr to only operate on the actual result set. So if you only
  search for 10 items, you get them really quick. If you search for 100000, it
  will still be slow. (pilz)


1.4.26 (2013-07-15)
-------------------

- Don't set 'rows' or 'start' in the SOLR fallback search if they
  already exist. [regebro]


1.4.25 (2013-06-28)
-------------------

- Use searchResults() not search() in the language fallback, to make sure
  permissions are considered [pysailor]


1.4.24 (2013-06-27)
-------------------

- remove dependency on osha.applicaitonform, dammit [pysailor]

1.4.23 (2013-06-26)
-------------------

- Fixed #7855 by making the fallback also ignore TypeErrors + not do
  the fallback think for resolveuid calls. [regebro]
- Added a shortcut for the solr language fall back search, when 
  search results are empty. [regebro]
- Made blog migration more robust [pysailor]
- Allow PressRelease in Collage in cleanup_content too #7803 [pysailor]


1.4.22 (2013-06-24)
-------------------

- Language fallback #7734
  [lregebro, reinhard, pysailor]
- Added ellipsis for teasers on homepage #7602 [deroiste]
- Comment out handle_object_willbe_translated, since we don't have LinguaLinks any more
  [pysailor]
- add collective.portlet.twitter to dependencies, refs #7781 [jure]


1.4.21 (2013-05-28)
-------------------

- #7648 Link checker: replace collective.lead with sqlalchemy for
   updating the postgres db [deroiste, pilz]


1.4.20 (2013-05-24)
-------------------

- Comment out db adapter, to restore the IDatabaseSettings config
  [deroiste, thomasw]

1.4.19 (2013-05-14)
-------------------

- #7358 set "use_email_as_login" in the portal site_properties. This
  configuration change has been manually applied to production
  already. [deroiste]
- #7376 re-enable the lc_maintenance view to fix the BIRT broken link
  reports [deroiste]
- #6914 js for seminarportal [jure]
- Don't include seminarportal.css directly in the speech_summary template to prevent KeyError: 'base_properties' (we're using css registry) #7490 [jure]


1.4.18 (2013-04-23)
-------------------

- LCRetrieveByPath: don't retireve content that is private [pysailor]


1.4.17 (2013-04-17)
-------------------

- helper browser view for cleaning up old content #7211 [pysailor]
- added align_languages action and view [reinhardt]

1.4.16 (2013-04-08)
-------------------

- Nothing changed yet.


1.4.15 (2013-04-08)
-------------------

- add osha.applicationform to dependencies, refs #6165 [jure]
- add test for EnableJSView, refs #6344 [jure]
- Add an upgrade step to clear the linkchecker database #6952 #6977 [deroiste]
- make registerManyLinks async. Fix #7032 [pilz]
- Add tests for translationhelper and fix bugs in the find_default method, refs #5152 [jure]
- #6980 only check links in public items which are not outdated [deroiste]
- #7038 register a link checker retriever for SPSpeaker and Topic
   items [deroiste]

1.4.14 (2013-02-15)
-------------------

- Add translation for "Privacy Policy" (footer) #5998 [deroiste]
- Change sender for ecards #6273 [thomasw] 

1.4.13 (2013-01-31)
-------------------

- change default output type to text/x-html-safe for news item description field
  to make ckeditor internal links work, refs #6554 [jure]

1.4.12 (2013-01-08)
-------------------

- When an oshmail is created and prefilled, news items are sorted in reversed order #6484 [thomasw]

1.4.11 (2013-01-02)
-------------------

- #5899 Fix pressrelease_view for HWC2010 [deroiste]


1.4.10 (2012-12-18)
-------------------

- Nothing changed yet.


1.4.9 (2012-12-18)
------------------


src/osha/policy/tools.py
#from lxml.etree import parse, Element, tostring
from elementtree.ElementTree import parse, Element, tostring
from xlrd import open_workbook
import csv
import logging
import sys

def get_xls(filename):
    book = open_workbook(file_contents=file(filename).read())
    sheet = book.sheets()[0]
    return sheet

class Command(object):
    def __init__(self, command_type, original_id = None, old_id = None,
                 new_move_id = None, create_id = None, caption = None,
                 description = None, parent_id = None):
        self.command_type = command_type
        self.original_id = original_id
        self.old_id = old_id
        self.new_move_id = new_move_id
        self.create_id = create_id
        self.caption = caption
        self.description = description
        self.parent_id = parent_id

    def __cmp__(self, other):
        our_level = (self.old_id and self.old_id[-1]) or (self.original_id and self.original_id[-1]) or (self.new_move_id and self.new_move_id[-1]) or (self.create_id and self.create_id[-1])
        their_level = (other.old_id and other.old_id[-1]) or (self.original_id and self.original_id[-1]) or (other.new_move_id and other.new_move_id[-1]) or (other.create_id and other.create_id[-1])
        if our_level != their_level:
            if our_level < their_level:
                return 1
            return -1
        else:
            for order in ('c', 'm', 'd', 'u'):
                if self.command_type == order:
                    return -1
                if other.command_type == order:
                    return 1

    def show_cmd(self):
        cmd, original_id, old_id, new_move_id, create_id, caption, description, parent_id = \
            self.command_type, self.original_id, self.old_id, self.new_move_id, self.create_id, self.caption, \
            self.description, self.parent_id
        if cmd == 'c':
            return '|'.join((cmd, create_id, parent_id, caption, description))
        if cmd == 'u':
            return '|'.join((cmd, original_id, caption, description))
        if cmd == 'm':
            return '|'.join((cmd, old_id, new_move_id, parent_id))
        if cmd == 'd':
            return '|'.join((cmd, old_id))

def converter():
    try:
        _converter()
    except Exception, e:
        print e
        import pdb;pdb.post_mortem(sys.exc_traceback)

def _converter():
    sheet = get_xls(sys.argv[1])

    commands = []
    for i in range(sheet.nrows):
        row = sheet.row(i)
        command = row[9].value.lower().strip()
        create_id = row[10].value
        old_id = row[13].value.strip()
        original_id = row[0].value.strip()
        new_move_id = row[14].value.strip()
        caption = row[11].value.strip()
        description = row[12].value.strip()
        parent_id = row[17].value.strip()
        args = (original_id, old_id, new_move_id, create_id, caption, description, parent_id)

        if not (command or old_id):
            continue
        if command == 'new term':
            if old_id:
                #Howdy, this is not a new term, this is a move and rename action
                commands.append(Command('u', original_id = new_move_id, caption = caption, description = description))
                commands.append(Command('m', original_id = original_id, old_id = old_id, new_move_id = new_move_id, caption = caption, parent_id = parent_id))
            else:
                commands.append(Command('c', create_id = create_id, parent_id = parent_id, caption = caption, description = description))
        elif command == 'amend term':
            commands.append(Command('u', original_id = original_id, caption = caption, description = description))
            if old_id:
                #Howdy, that is ALSO a hidden move. How sneaky!
                commands.append(Command('m', original_id = original_id, old_id = old_id, new_move_id = new_move_id, caption = caption, parent_id = parent_id))
        elif command == 'delete term' and not old_id:
            commands.append(Command('d', old_id = old_id))
        elif old_id and not command == u'accept / reject':
            commands.append(Command('m', old_id = old_id, new_move_id = new_move_id, caption = caption, parent_id = parent_id))
        elif command == u'accept / reject':
            continue
        else:
            raise Exception("Unknown case")

    cleaning = True
    while cleaning:
        cleaning = False
        id_to_command = {}
        for command in commands:
            if command.command_type == 'm':
                if not id_to_command.has_key(command.old_id):
                    id_to_command[command.old_id] = []
                id_to_command[command.old_id].append(command)
        for cmds in id_to_command.values():
            removes = set()
            for key in cmds[0].__dict__.keys():
                different_values = len(set([getattr(x, key) for x in cmds if getattr(x, key)]))
                if key == 'parent_id' and different_values > 1:
                    parents = [int(x.parent_id[:-1]) for x in cmds if x.parent_id]
                    new_id = int(cmds[0].new_move_id[:-1])
                    closest_parent = ''
                    closest = 99999999999999
                    for cmd in cmds:
                        parent_distance = cmd.parent_id and new_id - int(cmd.parent_id[:-1]) or 999999999999999
                        if parent_distance > 0 and closest > parent_distance:
                            closest_parent = cmd
                            closest = parent_distance
                    removes.update([x for x in cmds if x.parent_id != closest_parent.parent_id])
                    continue
                assert different_values <= 1, "Two move lines have different data. Key: '%s', values: '%s'" % (key, '|'.join([getattr(x, key) for x in cmds]))
                if different_values:
                    removes.update([x for x in cmds if not getattr(x, key)])
            if not cmds[0].old_id in ['61721D', '18201F']:
                assert (set(cmds) - removes).pop()
            for cmd in removes:
                commands.remove(cmd)
                cleaning = True
                break
    commands.sort()
    for command in commands:
        print command.show_cmd()


NS = '{http://www.imsglobal.org/xsd/imsvdex_v1p0}'

debugging_handler = logging.StreamHandler()
debugging_log = logging.getLogger('debugging_log')
debugging_log.addHandler(debugging_handler)
debugging_log.setLevel(logging.WARN)

class BaseCommand(object):
    def run(self):
        return NotImplemented()

    def untranslated_ids(self):
        return NotImplemented

class CreateCommand(BaseCommand):
    def __init__(self, tree, id, parent_id, caption, description):
        self.tree = tree
        self.id = id
        self.parent_id = parent_id
        self.caption = caption
        self.description = description

    def run(self):
        termIdentifier = self.id
        termCaption = self.caption
        termDescription = self.description
        assert termIdentifier not in self.tree.items.keys()
        parentTerm = self.tree.items[self.parent_id]
        new_term =  self.tree.new_tree_term(termIdentifier, termCaption, termDescription)
        last_brother = None
        has_brothers = False
        insert_check = False
        for index, brother in enumerate(parentTerm):
            if brother.tag != NS + 'term':
                continue
            else:
                has_brothers = True
            brotherTermIdentifier = brother[0].text
            if termIdentifier < brotherTermIdentifier:
                insert_check = True
                parentTerm.insert(index, new_term)
                last_brother = None
                break
            last_brother = index
        if last_brother:
            insert_check = True
            parentTerm.insert(last_brother + 1, new_term)
        if not has_brothers:
            insert_check = True
            parentTerm.insert(index, new_term)
        assert insert_check

    def untranslated_ids(self):
        return self.id, self.caption, self.description

class UpdateCommand(BaseCommand):
    def __init__(self, tree, id, caption, description):
        self.tree = tree
        self.id = id
        self.caption = caption
        self.description = description

    def run(self):
        term = self.tree.items[self.id]
        new_term = self.tree.new_tree_term('ignore', self.caption, self.description)
        caption = new_term[1]
        description = new_term[2]
        term.remove(term[1])
        term.insert(1, caption)
        for i in range(20):
            if term[i].tag == '{http://www.imsglobal.org/xsd/imsvdex_v1p0}description':
                break
        term.remove(term[i])
        term.insert(i, description)

    def untranslated_ids(self):
        return self.id, self.caption, self.description

class MoveCommand(BaseCommand):
    def __init__(self, tree,  old, new, new_father, keep_names):
        self.keep_names = 'keep_names' == keep_names.lower()
        self.tree = tree
        self.old = old
        self.new = new
        self.new_father = new_father

    def run(self):
        old_child = self.tree.items[self.old]
        old_father = self.tree.fathers[old_child]
        old_father.remove(old_child)
        new = self.new
        if new in self.tree.items.keys():
            if not self.keep_names:
                new_term = self.tree.items[new]
                caption = old_child[1]
                description = old_child[2]
                new_term.remove(new_term[1])
                new_term.insert(1, caption)
                for i in range(20):
                    if new_term[i].tag == '{http://www.imsglobal.org/xsd/imsvdex_v1p0}description':
                        break
                new_term.remove(new_term[i])
                new_term.insert(i, description)

            return
        parentTerm = self.tree.items[self.new_father]
        last_brother = None
        has_brothers = False
        for index, brother in enumerate(parentTerm):
            if brother.tag != NS + 'term':
                continue
            else:
                has_brothers = True
            brotherTermIdentifier = brother[0].text
            if new < brotherTermIdentifier:
                parentTerm.insert(index, old_child)
                last_brother = None
                break
            last_brother = index
        if last_brother:
            parentTerm.insert(last_brother + 1, old_child)
        if not has_brothers:
            parentTerm.insert(index, old_child)

        old_child[0].text = new

    def untranslated_ids(self):
        return None

class DeleteCommand(BaseCommand):
    def __init__(self, tree, id):
        self.tree = tree
        self.id = id

    def run(self):
        kid_id = self.id
        kid = self.tree.items[kid_id]
        father = self.tree.fathers[kid]
        father.remove(kid)
        return False

    def untranslated_ids(self):
        return None

class Tree(object):
    def __init__(self, thesaurus_file):
        self.__tree = parse(thesaurus_file)
        self.all_languages = set([x.get('language') for x in self.__tree.findall('.//%slangstring' % NS)])
        self.update()

    def update(self):
        self.fathers = {}
        self.items = {}
        for term in self.__tree.findall('.//%sterm' % NS):
            self.items[term[0].text] = term
        for term in self.__tree.findall('//%sterm' % NS):
            for child in term:
                self.fathers[child] = term

    def new_tree_term(self, termIdentifier, caption, description):
        term = Element(NS + 'term')
        treeTermIdentifier = Element(NS + 'termIdentifier')
        treeTermIdentifier.text = termIdentifier
        term.append(treeTermIdentifier)
        treeCaption = Element(NS + 'caption')
        self.addLangs(treeCaption, caption)
        term.append(treeCaption)
        treeDesc = Element('description')
        self.addLangs(treeDesc, description)
        term.append(treeDesc)
        return term

    def addLangs(self, element, msg):
        for lang in self.all_languages:
            elem = Element(NS + 'langstring')
            elem.set('language', lang)
            elem.text = msg
            element.append(elem)

    def close(self):
        self.__tree.write('out.xml', 'utf-8')

def cmd_parser(line, tree):
    tokens = line.strip().split('|')
    cmd_token = tokens[0]
    args = tokens[1:]
    cmd = {'c' : CreateCommand,
           'u' : UpdateCommand,
           'm' : MoveCommand,
           'd' : DeleteCommand}[cmd_token](tree, *args)
    return cmd

def vdexupdater():
    try:
        _vdexupdater()
    except Exception, e:
        print e
        import pdb;pdb.post_mortem(sys.exc_traceback)


def _vdexupdater():
    if len(sys.argv) > 2:
        thesaurus_file = sys.argv[2]
    else:
        thesaurus_file = './src/osha/policy/data/vocabularies/MultilingualThesaurus.vdex'
    tree = Tree(thesaurus_file)
    translations = []
    for line in file(sys.argv[1]):
        print line
        #print '23064E' in tostring(tree._Tree__tree.getroot())
        instance = cmd_parser(line, tree)
        instance.run()
        translations.append(instance.untranslated_ids())
        tree.update()
    tree.close()

def xmltoxls():
    try:
        _xmltoxls()
    except Exception, e:
        print e
        import pdb;pdb.post_mortem(sys.exc_traceback)

def _xmltoxls():
    if len(sys.argv) > 1:
        xml = parse(sys.argv[1])
    else:
        xml = parse('out.xml')
    out = csv.DictWriter(file('out.csv', 'w'), ['id', 'levela', 'levelb', 'levelc', 'leveld', 'levele', 'levelf', 'levelg', 'levelh', 'translations'])
    for term in xml.findall('.//%sterm' % NS):
        id = term.find('%stermIdentifier' % NS).text
        level = 'level' + id[-1].lower()
        caption = [x.text for x in term.findall('%scaption/%slangstring' % (NS, NS)) if x.attrib['language'] == 'en'][0].encode('ascii', 'ignore')
        translations = len(set([x.text for x in term.findall('%scaption/%slangstring' % (NS, NS))]))
        new_row = {}
        new_row['id'] = id
        new_row[level] = caption
        new_row['translations'] = translations
        print id, new_row.keys()
        out.writerow(new_row)

def ordervdex():
    try:
        _ordervdex()
    except Exception, e:
        print e
        import pdb;pdb.post_mortem(sys.exc_traceback)

def _ordervdex():
    if len(sys.argv) > 1:
        vdex_filename = sys.argv[1]
    else:
        vdex_filename = 'out.xml'
    vdex = parse(vdex_filename)
    orderNode(vdex.getroot())
    vdex.write(vdex_filename, 'utf-8')

def orderNode(node):
    terms = [(index, x) for (index, x) in enumerate(node) if x.tag.endswith('term')]
    current_order = [x[0] for x in terms]
    map(orderNode, [x[1] for x in terms])
    sorter = lambda a, b: unicode(a[1][0].text.strip()).__cmp__(unicode(b[1][0].text.strip()))
    terms.sort(sorter)
    new_order = [x[0] for x in terms]
    if new_order != current_order:
        print "%s has unsorted children" % node[0].text
        for ((ignore, term), new_pos) in zip(terms, current_order):
            node.remove(term)
            node.insert(new_pos, term)

def find_missing_translations():
    try:
        _find_missing_translations()
    except Exception, e:
        print e
        import pdb;pdb.post_mortem(sys.exc_traceback)

def _find_missing_translations():
    levels = [None for i in range(6)]
    if len(sys.argv) > 1:
        filename = sys.argv[1]
    else:
        filename = 'out.csv'
    data = csv.reader(file(filename))
    translations = ['id', 'Path (for context only)', 'Term to translate','translation']
    writer = csv.writer(open('translate.csv', 'wb'))
    writer.writerow(translations)
    for row in data:
        translations = int(row[9])
        for level in range(1, 7):
            if row[level]:
                levels[level-1] = row[level]
                break
        if translations > 1:
            continue

        id = row[0]
        ttt = row[level]
        path = ' -> '.join(levels[:level])
        writer.writerow([id, path, ttt, ''])

def insertTranslations():
    try:
        _insertTranslations()
    except Exception, e:
        print e
        import pdb;pdb.post_mortem(sys.exc_traceback)

def _insertTranslations():
    xml = file(sys.argv[1])
    vdex = parse(xml)
    root = vdex.getroot()
    terms = {}
    keys = {}
    for term in root.findall('.//%sterm' % NS):
        terms[term[0].text.lower()] = term
    for translation_filename in sys.argv[2:]:
        language = translation_filename[27:29].lower()
        xls = get_xls(translation_filename)
        keys[language] = []
        for i in range(1, xls.nrows):
            row = xls.row(i)
            id = row[0].value.lower().strip().lower()
            keys[language].append(id)
            translation = row[3].value.strip()
            assert translation
            term = terms[id]
            found = False
            for language_term in term.findall('./%scaption/%slangstring' % (NS, NS)):
                if language_term.attrib['language'] == language:
                    language_term.text = translation
                    if found:
                        assert False, "TWICE???"
                    found = True
            assert found
    ref_ids = keys.values()[0]
    for ids in keys.values():
        assert ref_ids == ids
    vdex.write('new.vdex', 'utf-8')
            
